### PR TITLE
ERC-677 transferAndCall

### LINKED ISSUE #17: ERC677: transferAndCall
Any thoughts on including transferAndCall? Should hopefully make some transactions cheaper by not needing to use allowances.

https://github.com/ethereum/EIPs/issues/677

### PR DESCRIPTION
**Pull request by dmihal**
Fixes #17

### CODE REVIEW

**Comment by albertocuestacanada** [comment]
Give me until Tuesday or Wednesday to review this, I want to double check all methods for reentrancy vulnerabilities.

I know that `withdraw` already opens the same opportunities for reentrancy attacks, but just anyway.

**Comment by albertocuestacanada** [comment]
@dmihal, could you please add tests for this?

File: contracts/ERC677Receiver.sol [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.6.10;
+
+interface ERC677Receiver {
```
Comment (albertocuestacanada) [inline contracts/ERC677Receiver.sol:4]: Do we want the interfaces in separate files or in `WETH10.sol`.

I'm in favour that when they are one-liners we should follow the MakerDAO pattern and add them in `WETH.sol`.

File: contracts/ERC677Receiver.sol [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+    function onTokenTransfer(address _sender, uint _value, bytes memory _data) external;
```
Comment (albertocuestacanada) [inline contracts/ERC677Receiver.sol:5]: Please remove the leading underscore from parameters, it adds unnecessary noise, in my opinion.

File: contracts/WETH10.sol [hunk 3]
```diff
@@ -85,6 +88,15 @@ contract WETH10 {
 
         return true;
     }
+
+    function transferAndCall(address _to, uint _value, bytes memory _data) public returns (bool success) {
```
Comment (albertocuestacanada) [inline contracts/WETH10.sol:92]: Please remove leading underscores.

Comment (albertocuestacanada) [inline contracts/WETH10.sol:92]: Would it be cheaper to use `bytes calldata data`?

File: contracts/WETH10.sol [hunk 4]
```diff
@@ -85,6 +88,15 @@ contract WETH10 {
...
+        transferFrom(msg.sender, _to, _value);
+
+        if (isContract(_to)) {
```
Comment (albertocuestacanada) [inline contracts/WETH10.sol:95]: Does it make any sense to check if the receiver is a contract or an EOA?

Even if the receiver is a contract, there is no guarantee that with implement ERC667Receiver, so we might as well skip this check.

File: contracts/ERC677Receiver.sol (hunk 1 above)
Comment (dmihal) [reply contracts/ERC677Receiver.sol]: I've moved it to the top of WETH.sol

**Comment by dmihal** [comment]
Comments addressed & tests added :)

**Review by albertocuestacanada — APPROVED** [review]
It was a bit of a messy merge, but I think I got it right.

After the merge, we might want to consider:
 - Consistent interface naming.
 - Make `transferFrom` external and copy the code inside `transferAndCall`.

**Pull request merged**